tool-openssl/x509.cc

```diff
@@ -3,72 +3,251 @@
       return false;
     }
 
-    // Parse x509 certificate from input file
-    const uint8_t *p = input_data.data();
-    bssl::UniquePtr<X509> x509(d2i_X509(nullptr, &p, input_data.size()));
+    // Set validity period, default 30 days if not specified
+    int valid_days = days > 0 ? days : 30;
+    if (!X509_gmtime_adj(X509_getm_notBefore(x509.get()), 0) ||
+      !X509_gmtime_adj(X509_getm_notAfter(x509.get()), 60 * 60 * 24 * static_cast<long>(valid_days))) {
+      fprintf(stderr, "Error: unable to set validity period\n");
+      return false;
+    }
+
+    // Sign the certificate with the provided key
+    if (!signkey_path.empty()) {
+      if (!LoadAndSignCertificate(x509.get(), signkey_path)) {
+        return false;
+      }
+    }
+
+    // Write the signed certificate to output file
+    if (!noout && !out_path.empty()) {
+      if (!WriteSignedCertificate(x509.get(), out_path)) {
+        return false;
+      }
+    }
+  } else {
+    // Parse x509 certificate from input PEM file
+    bssl::UniquePtr<X509> x509(PEM_read_X509(in_file.get(), nullptr, nullptr, nullptr));
     if (!x509) {
-        fprintf(stderr, "Failed to parse X509 certificate from '%s'.\n", in_path.c_str());
-        ERR_print_errors_fp(stderr);
+      fprintf(stderr, "Error: error parsing certificate from '%s'\n", in_path.c_str());
+      ERR_print_errors_fp(stderr);
+      return false;
+    }
+
+    if (modulus) {
+      bssl::UniquePtr<EVP_PKEY> pkey(X509_get_pubkey(x509.get()));
+      if (!pkey) {
+        fprintf(stderr, "Error: unable to load public key from certificate\n");
         return false;
+      }
+
+      if (EVP_PKEY_base_id(pkey.get()) == EVP_PKEY_RSA) {
+        const RSA *rsa = EVP_PKEY_get0_RSA(pkey.get());
+        if (!rsa) {
+          fprintf(stderr, "Error: unable to load RSA key\n");
+          return false;
+        }
+        const BIGNUM *n = RSA_get0_n(rsa);
+        if (!n) {
+          fprintf(stderr, "Error: unable to load modulus\n");
+          return false;
+        }
+        printf("Modulus=");
+        BN_print_fp(stdout, n);
+        printf("\n");
+      } else {
+        fprintf(stderr, "Error: public key is not an RSA key\n");
+      }
     }
 
-    // Serialize certificate to DER format
-    uint8_t *out_data = nullptr;
-    int len = i2d_X509(x509.get(), &out_data);
-    if (len < 0) {
-        fprintf(stderr, "Failed to serialize X509 certificate.\n");
-        ERR_print_errors_fp(stderr);
+    if (dates) {
+      char not_before[20], not_after[20];
+      ASN1_TIME *nb = X509_getm_notBefore(x509.get());
+      ASN1_TIME *na = X509_getm_notAfter(x509.get());
+      struct tm tm;
+      ASN1_TIME_to_tm(nb, &tm);
+      strftime(not_before, sizeof(not_before), "%Y-%m-%d %H:%M:%S", &tm);
+      ASN1_TIME_to_tm(na, &tm);
+      strftime(not_after, sizeof(not_after), "%Y-%m-%d %H:%M:%S", &tm);
+      printf("notBefore=%s\n", not_before);
+      printf("notAfter=%s\n", not_after);
```

Comment:
Slight nit: It would be best to let the `printf` and `fprintf` calls be consistent.

Reply:
updating this and dates output to match OpenSSL output